### DISCUSSION TITLE
Primary Key for storing raw messages in database

### DISCUSSION THREAD

**Discussion by ghost** [opendcs/opendcs#286] [Q&A]
I am storing raw DCP messages in a database.  The DDL looks like

```
CREATE TABLE `goes_raw_data` (
  `message` varchar(100) NOT NULL,
  `goes_id` varchar(20) NOT NULL,
  `datetime` datetime NOT NULL,
  `failure_code` varchar(1) NOT NULL,
  `signal_strength` int(11) NOT NULL,
  `frequency_offset` int(11) NOT NULL,
  `modulation_index` varchar(1) NOT NULL,
  `data_quality_indicator` varchar(1) NOT NULL,
  `goes_receive_channel` int(11) NOT NULL,
  `goes_spacecraft_indicator` varchar(1) NOT NULL,
  `data_source_code` varchar(2) NOT NULL,
  `message_data_length` int(11) NOT NULL,
  UNIQUE KEY `goes_raw_data_UN` (`message`)
) ENGINE=InnoDB AUTO_INCREMENT=26007 DEFAULT CHARSET=latin1;
```

Will the combination of  `goes_id`, `datetime`, and `goes_receive_channel` will always be unique? Basically, is it safe to make PRIMARY KEY out of `goes_id`, `datetime`, and `goes_receive_channel`?

**Comment by MikeNeilson** [comment]
I would add space craft and failure code (unless you only save good messages.)

Also let me know if you are interesting in using the other pieces of OpenDCS with your custom database ( you can look in the "schema/opendcs-pg" folder. In theory there's already support but to have an actual outside implementation will require some changes I'm planning to make; knowing someone else wants to use it that way will help motivate me and give me a reason to tweak my timelines a little depending on exactly what you need when.

You might also look at using the RoutingSpecs/RoutingScheduler and creating a "Consumer" there in.

**Reply by MikeNeilson** [reply]
Oh, and data source code. That would identify different LRGS system... I think, that might actually be specific to how we have the LRGS setup.

**Reply by ghost** [reply]
Sounds like I should use the whole header (first 37 bytes) as the primary key

**Reply by MikeNeilson** [reply]
Hmm, yeah. though I would still list them as separate values. Then again if you wanted to handle more than GOES (the LRGS can do Irridum and faked GOES over ethernet) some sort of encoded key might be the way to go.

really depends on what kind of analysis/lookup if any you'd want to do.
